### PR TITLE
Fungible Asset Standard

### PR DESCRIPTION
**Pull request by xlc**
A fungible asset standard for Substrate-based blockchains

To keep the initial version reasonable scoped, only core features are considered.

In scope:

- Query balance
- In-chain transfer asset

Out of scope:

- Reserve asset
- Lock asset
- Mint/Burn asset
- Cross-chain interactions
- Asset metadata registry

### CODE REVIEW

**Comment by gavofyork** [comment]
Looks reasonable enough as a document that essentially describes the effective interface of `balances` and `assets` modules. That said, storage items are rather low-level things to be placing in a spec and I can easily imagine future modules doing anything non-trivial will want to move away from the storage format of these early modules. As such, we will want to introduce some means of expressing balances, reservations and illiquidity in a way that provides flexibility for how such information is represented on-chain. I expect this can be done through exporting wasm blobs which hold logic for querying the storage items and perhaps events and creating transactions in order to do this manipulation through a standard API.

**Comment by jak-pan** [comment]
Can we open discussion on this? 

I guess that **[this](https://github.com/open-web3-stack/open-runtime-module-library/tree/master/tokens)** is the POC implementation of this standard and it looks good to me. We'll be switching to this implementation and suggest additions if needed for our use case in the coming days.

**Comment by xlc** [comment]
I am starting rewriting this to reflect the latest development of Substrate & Polkadot.

**Comment by burdges** [comment]
I suspect the biggest issue is the word standard in the title.  We've multiple goals that should radically alter how transactions are handled and store balances, which currently remain blocked blocked on implementing parachians.

We won't aggressively monkey with how assets are staked however because anonymous staking gets messy, so maybe something like "Easy Stakable Fungible Asset Interface".

**Comment by lucasvo** [comment]
Are you trying to cover both the functionality of native tokens such as DOTs or anything that uses balances? I think the power of the ERC20 standard is that it is extremely simple. I would propose focusing solely on a spec for assets and avoid covering any additional functionality that balances have (the notion of locked/free/reserved balances for example). 

I think roughly the following features that are generally associated with fungible assets should be discussed:
* State rent / minimum deposits
* Basic transfer, totalSupply and balance interface
* Mint/Burn
* Identification of tokens (on Ethereum addresses)
* Metadata

I see there are comments about staking in here as well, does this belong in a basic fungible asset spec?

Do you have a public draft? We (centrifuge) would love to contribute to it. Perhaps we can start off with a version of it on hackmd and edit it together?

**Comment by jak-pan** [comment]
I think that recently released XCM format draft spec will be a good starting point and we could iterate on that. It would be great if cross chain assets had the same features and interfaces as the native asset has (transfers, locking, staking etc.). 
To start, we need a way to generate unique asset ID which XCM draft lightly touches and will probably need to be coordinated with that.

**Comment by xlc** [comment]
@burdges I am not sure if I get your points fully but this aims to define an universal interface / standard that should be implemented by everyone in the ecosystem to allow simplified integration with wallet / exchange / dapp / cross-chain interaction. It should be flexible enough to support all different kinds of implementations.

@lucasvo Yes this aims to cover both the native tokens as well as custom tokens. I agree at least for initial version we should focus on core features first.

> I think roughly the following features that are generally associated with fungible assets should be discussed:
> 
> * State rent / minimum deposits

Not every chain is going to have state rent / minimum deposits. So this should allow implementation to have such feature without actual defining it.

> * Basic transfer, totalSupply and balance interface

Yes

> * Mint/Burn

Again not every tokens supports this. I would like to leave it as a future extension.

> * Identification of tokens (on Ethereum addresses)
> * Metadata

**Comment by jak-pan** [comment]
> while this is designed to towards to enable cross chain interaction, I would like to keep this as simple as possible just so we can get something done first. So my initial focus will be just within a single chain which goal in mind to build a solid foundation for cross chain interaction.

@xlc I'm fine with that approach. What I really wanted is the same API for interacting with native asset balance and non-native assets to not unnecessarily complicate the code when interacting with multiple assets. It looks good to me RN.
One question though. Do we need both total & free balance if we are not locking?

**Comment by xlc** [comment]
> Do we need both total & free balance if we are not locking?

I don't have a strong opinions on this. I was thinking to kind of support it without support it. But wasn't very sure if this is the best way.

File: PSPs/psp-fungible-asset-standard.md [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+  };
+  events: {
+  	[name: string]: {
+      types: Record<string, TypeName>;
+      path: Path;
+    } 
+  };
+}
+
+interface TypeWithHahser {
```
Comment (shaopengw) [inline PSPs/psp-fungible-asset-standard.md:88]:
```suggestion
interface TypeWithHasher {
```

File: PSPs/psp-fungible-asset-standard.md [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+  name: string;
+  symbol: string;
+  decimals: number;
+  uri?: string; // Substrate URI. See psp-uri
+  description?: string;
+  icon?: string; // HTTP URL or IPFS hash or dataURI
+}
+```
+
+Asset description can also be embedded in genesis file as additional properties. It is expected to be immutable but maybe changed under speical circumstances (e.g. project rename / rebrand).
```
Comment (shaopengw) [inline PSPs/psp-fungible-asset-standard.md:56]:
```suggestion
Asset description can also be embedded in genesis file as additional properties. It is expected to be immutable but maybe changed under special circumstances (e.g. project rename / rebrand).
```

File: PSPs/psp-fungible-asset-standard.md [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+  type: TypeName;
+  hasher: Hasher;
+}
+type TypeName = string;
+type Hasher = 'Blake2_128' | 'Blake2_256' | 'Blake2_128Concat' | 'Twox128' | 'Twox256' | 'Twox64Concat' | 'Identity';
+type Hasher = ''
+type Path = string | [ModuleName, MethodName, ...string[]]
+```
+
+Asset operation description stores the nessusary information to perform asset operations and might need to be updated with new runtime versions. There should be a way to fetch the corresponding `AssetOperationDescription` . e.g. From a registry or a commonly known location.
```
Comment (shaopengw) [inline PSPs/psp-fungible-asset-standard.md:98]:
```suggestion
Asset operation description stores the necessary information to perform asset operations and might need to be updated with new runtime versions. There should be a way to fetch the corresponding `AssetOperationDescription` . e.g. From a registry or a commonly known location.
```

File: PSPs/psp-fungible-asset-standard.md [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+
+#### TypeRegistry
+
+Type registry stores the nessusary type information to interact with the runtime. This is inspired by polkadot.js type registry.
```
Comment (shaopengw) [inline PSPs/psp-fungible-asset-standard.md:102]:
```suggestion
Type registry stores the necessary type information to interact with the runtime. This is inspired by polkadot.js type registry.
```

File: PSPs/psp-fungible-asset-standard.md [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+  };
+}
+
+interface TypeWithHahser {
+  type: TypeName;
+  hasher: Hasher;
+}
+type TypeName = string;
+type Hasher = 'Blake2_128' | 'Blake2_256' | 'Blake2_128Concat' | 'Twox128' | 'Twox256' | 'Twox64Concat' | 'Identity';
+type Hasher = ''
```
Comment (shaopengw) [inline PSPs/psp-fungible-asset-standard.md:94]: Looks like duplicated `Hasher` definition here.

File: PSPs/psp-fungible-asset-standard.md [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+#### Data Types
+
+With TypeScript syntax:
+
+```typescript
+interface AssetOperationDescription {
+  types: TypeRegistry;
+  queries: {
+    [name: string]: {
+      parameterTypes?: Record<string, TypeWithHahser>;
```
Comment (wischli) [inline PSPs/psp-fungible-asset-standard.md:69]:
```
      parameterTypes?: Record<string, TypeWithHasher>;
```

File: PSPs/psp-fungible-asset-standard.md [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+
+Note: Substrate Runtime Metadata is required to encode / decode dispatchable calls and events. The details of Substrate Runtime Metadata is out of scope.
+
+#### TypeRegistry
+
+Type registry stores the necessary type information to interact with the runtime. This is inspired by polkadot.js type registry.
+
+```typescript
+interface TypeRegistry {
+  [typename: string]: TypeDefination;
```
Comment (wischli) [inline PSPs/psp-fungible-asset-standard.md:107]: I assume you meant to say `TypeDefinition`?

Comment (xlc) [reply PSPs/psp-fungible-asset-standard.md]: oops, apparently I am really bad at spelling...

**Comment by danforbes** [comment]
Very minor nitpick but I would suggest using a 100-character maximum line length for all documents of this type.

**Comment by lucasvo** [comment]
@xlc do you think we could use a more generic example that does not refer to acala's tokens?

**Comment by xlc** [comment]
@lucasvo 

I used Acala as example for few reasons:

- I want to avoid hypothetical example because that often leads to wrong path
- It is easier to come out an example by looking at some existing code
- I intent to make Acala as an reference implementation

But I can just search & replace `Acala` to `Example` if people believe it is more appropriate

**Comment by Noc2** [comment]
Merged as discussed on riot.

**Pull request merged**